Expanding CircuitBreakerMapping to include request path as a key

haithamgabr opened the pull request:
This PR introduces request path as a third dimension to the key of CircuitBreakerMapping.

Feature request issue: https://github.com/line/armeria/issues/3134

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java:
```diff
@@ -183,6 +212,95 @@
+     *
+     * <p>Since {@link CircuitBreaker} is a unit of failure detection, don't reuse the same instance for
+     * unrelated services.
+     *
+     * @param factory a function that takes a method+path and creates a new {@link CircuitBreaker}.
+     */
+    public static Function<? super HttpClient, CircuitBreakerClient>
+    newPerMethodAndPathDecorator(BiFunction<String, String, ? extends CircuitBreaker> factory,
+                                 CircuitBreakerRuleWithContent<HttpResponse> ruleWithContent) {
+        return newDecorator(CircuitBreakerMapping.perMethodAndPath(factory), ruleWithContent);
+    }
+
+    /**
+     * Creates a new decorator that binds one {@link CircuitBreaker} per host, request method and path with
+     * the specified {@link CircuitBreakerRule}.
+     *
+     * <p>Since {@link CircuitBreaker} is a unit of failure detection, don't reuse the same instance for
+     * unrelated services.
+     *
+     * @param factory a function that takes a host+method+path and creates a new {@link CircuitBreaker}.
+     */
+    public static Function<? super HttpClient, CircuitBreakerClient>
+    newPerHostAndMethodAndPathDecorator(CircuitBreakerFactory factory, CircuitBreakerRule rule) {
+        return newDecorator(CircuitBreakerMapping.perHostAndMethodAndPath(factory), rule);
+    }
+
+    /**
+     * Creates a new decorator that binds one {@link CircuitBreaker} per host, request method and path with
+     * the specified {@link CircuitBreakerRuleWithContent}.
+     *
+     * <p>Since {@link CircuitBreaker} is a unit of failure detection, don't reuse the same instance for
+     * unrelated services.
+     *
+     * @param factory a function that takes a host+method+path and creates a new {@link CircuitBreaker}.
+     */
+    public static Function<? super HttpClient, CircuitBreakerClient>
+    newPerHostAndMethodAndPathDecorator(CircuitBreakerFactory factory,
+                                        CircuitBreakerRuleWithContent<HttpResponse> ruleWithContent) {
+        return newDecorator(CircuitBreakerMapping.perHostAndMethodAndPath(factory), ruleWithContent);
+    }
```
The factory methods in this class are overly crowded.
I think we can introduce a builder class that builds a `CircuitBreakerMapping` fluently.
For example:
```java
CircuitBreakerMapping mapping = 
    CircuitBreakerMapping.builder()
                         .perMethod()
                         .perPath()
                         ...
                        .build()

CircuitBreakerClient.newDecorator(mapping, rule);
```

haithamgabr replied:
Done.

haithamgabr commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/KeyedCircuitBreakerMapping.java:
```diff
@@ -57,20 +57,47 @@ public CircuitBreaker get(ClientRequestContext ctx, Request req) throws Exceptio
+                host = host(ctx);
+                method = method(ctx);
+                path = path(ctx);
+                key = host + '#' + method + "#" + path;
```
This part has now changed. I made sure to use the char.

haithamgabr commented:
I added CircuitBreakeMapping.Builder, left only the new perPath method, removed the other new double and triple methods, and deprecated the existing double method perHostAndMethod in favor of the new builder pattern.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMapping.java:
```diff
@@ -31,6 +30,62 @@
+    /**
+     * Builder class for building a CircuitBreakerMapping based on a combination of host, method and path.
+     */
+    class Builder {
```
- Could you move this builder class to top-level for keeping consistency with other builders?
- Add `static builder()` method for creating this builder?

haithamgabr replied:
Done

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMapping.java:
```diff
@@ -31,6 +30,62 @@
+         * @return a {@link CircuitBreakerMapping} based on the added dimensions.
+         */
+        public CircuitBreakerMapping build(CircuitBreakerFactory factory) {
+            return new KeyedCircuitBreakerMapping(this, factory);
```
How about raising an `IllegalStateException` if all `per{Host,Method,Path}` are `false`?

haithamgabr replied:
Done

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/KeyedCircuitBreakerMapping.java:
```diff
@@ -20,66 +20,59 @@
+        final String method = isPerMethod ? method(ctx) : null;
+        final String path = isPerPath ? path(ctx) : null;
+        final String key =
+                (isPerHost ? host : "") + '#' + (isPerMethod ? method : "") + '#' + (isPerPath ? path : "");
```
I think we can drop `#` if one of `isPer*` flags is false.
Guava `Joiner` could be useful for this case.
```java
static Joiner keyJoiner = Joiner.on('#').skipNulls();
String key = joiner.join(host, method, path);
```

haithamgabr replied:
Done through java streams.

haithamgabr commented:
I extracted CircuitBreakerMappingBuilder into a top-level class, and addressed the rest of comments.

trustin commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java:
```diff
@@ -161,7 +190,10 @@
      * unrelated services.
      *
      * @param factory a function that takes a host+method and creates a new {@link CircuitBreaker}.
+     *
+     * @deprecated Use newDecorator(), building a CircuitBreakerMapping using CircuitBreakerMapping.Builder().
      */
+    @Deprecated
```
Thanks! 🙇

haithamgabr replied:
My pleasure :)

trustin commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerFactory.java:
```diff
@@ -0,0 +1,35 @@
+@FunctionalInterface
+interface CircuitBreakerFactory {
+    /**
+     * Given a combination of host, method and path, creates a CircuitBreaker.
```
```suggestion
     * Given a combination of host, method and path, creates a {@link CircuitBreaker}.
```

trustin commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerFactory.java:
```diff
@@ -0,0 +1,35 @@
+     * @param host the host of the context endpoint.
+     * @param method the method of the context request.
+     * @param path the path of the context request.
+     * @return the CircuitBreaker instance corresponding to this combination.
```
```suggestion
     * @return the {@link CircuitBreaker} instance corresponding to this combination.
```

trustin commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerFactory.java:
```diff
@@ -0,0 +1,35 @@
+
+/**
+ * A functional interface that represents a mapper factory, mapping a combination of host, method and path
+ * to a CircuitBreaker.
```
```suggestion
 * to a {@link CircuitBreaker}.
```

trustin commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMapping.java:
```diff
@@ -22,30 +22,35 @@
     }
 
+    /**
+     * Returns a Builder that builds a CircuitBreakerMapping by setting host, method and/or path.
```
```suggestion
     * Returns a builder that builds a {@link CircuitBreakerMapping} by setting host, method and/or path.
```

trustin commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMappingBuilder.java:
```diff
@@ -0,0 +1,76 @@
+package com.linecorp.armeria.client.circuitbreaker;
+
+/**
+ * Builder class for building a CircuitBreakerMapping based on a combination of host, method and path.
```
```suggestion
 * Builder class for building a {@link CircuitBreakerMapping} based on a combination of host, method and path.
```

trustin commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMappingBuilder.java:
```diff
@@ -0,0 +1,76 @@
+
+    /**
+     * Adds host dimension to the mapping Key.
+     * @return this Builder.
```
```suggestion
     * @return this builder.
```

trustin commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMappingBuilder.java:
```diff
@@ -0,0 +1,76 @@
+
+    /**
+     * Adds method dimension to the mapping Key.
+     * @return this Builder.
```
```suggestion
     * @return this builder.
```

trustin commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMappingBuilder.java:
```diff
@@ -0,0 +1,76 @@
+
+    /**
+     * Adds path dimension to the mapping Key.
+     * @return this Builder.
```
```suggestion
     * @return this builder.
```

trustin reviewed (commented):
Nice work! Just left some minor comments.

haithamgabr commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMappingBuilder.java:
```diff
@@ -0,0 +1,76 @@
+    private boolean perHost;
+    private boolean perMethod;
+    private boolean perPath;
+
```
Done

haithamgabr commented:
I addressed @trustin's comments.

trustin commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMappingBuilder.java:
```diff
@@ -0,0 +1,78 @@
+        return perPath;
+    }
+
+    CircuitBreakerMappingBuilder() {}
```
Could be private and moved right after the member fields?

haithamgabr replied:
Moved it to the top, but has to stay package-private since it's called from CircuitBreakerMapping.builder()

trustin reviewed (approved):
Nice! Left one tiny nit. 🙇

minwoox commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMappingBuilder.java:
```diff
@@ -0,0 +1,78 @@
+
+    /**
+     * Builds the {@link CircuitBreakerMapping} using a three-dimensional factory.
+     * @return a {@link CircuitBreakerMapping} based on the added dimensions.
```
How about just:
```java
Returns a newly-created {@link CircuitBreakerMapping} with the specified {@link CircuitBreakerFactory} and properties set so far.
```

haithamgabr replied:
Done.

minwoox reviewed (commented):
Thanks! just left some minor comments.

haithamgabr commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerFactory.java:
```diff
@@ -0,0 +1,35 @@
+     * @param path the path of the context request.
+     * @return the {@link CircuitBreaker} instance corresponding to this combination.
+     */
+    CircuitBreaker apply(@Nullable String host, @Nullable String method, @Nullable String path);
```
I am not sure I understand what you're suggesting. They are nullable because they can be absent. e.g. map by host only, or host and method, or host and path, or all three.. etc.

minwoox replied:
Ah, I missed that we create the key as a combination of host, method and path.
So I think this is good as it is. 😄

minwoox commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMappingBuilder.java:
```diff
@@ -0,0 +1,78 @@
+        if (!perHost && !perMethod && !perPath) {
+            throw new IllegalStateException("A CircuitBreakerMapping must be per host, method and/or path");
+        }
+        return new KeyedCircuitBreakerMapping(this, factory);
```
How about just passing `perHost`, `perMethod` and `perPath`?
The builder is still mutable after `CircuitBreakerMapping` is created:
```java
CircuitBreakerMappingBuilder builder = ...
builder.build(factory);
// client makes a call
builder.perPath() // This change the behavior.
```

haithamgabr replied:
Done.

haithamgabr commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/KeyedCircuitBreakerMapping.java:
```diff
@@ -17,69 +17,66 @@
-    KeyedCircuitBreakerMapping(MappingKey mappingKey,
-                               BiFunction<String, String, ? extends CircuitBreaker> factory) {
-        this.mappingKey = requireNonNull(mappingKey, "mappingKey");
+    KeyedCircuitBreakerMapping(CircuitBreakerMappingBuilder mappingBuilder, CircuitBreakerFactory factory) {
```
This is actually the concrete implementation of CircuitBreakerMapping, so it cannot have itself as a member.

minwoox replied:
Yeah, I forgot to delete this comment after I left https://github.com/line/armeria/pull/3135#discussion_r516418946

haithamgabr commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMappingBuilder.java:
```diff
@@ -0,0 +1,78 @@
+
+    /**
+     * Adds host dimension to the mapping Key.
+     * @return this builder.
```
Done

haithamgabr commented:
Addressed @minwoox's comments.

minwoox commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/KeyedCircuitBreakerMapping.java:
```diff
@@ -17,69 +17,65 @@
-            default:
-                // should never reach here.
-                throw new Error();
+        if (!isPerHost && !isPerMethod && !isPerPath) {
```
We can check this only once in the constructor.
But we already check this in `build()` method so I think we can completely remove this. 😄

haithamgabr replied:
Done.

minwoox reviewed (approved):
Left one nit. Thanks a lot!

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerMappingBuilder.java:
```diff
@@ -0,0 +1,75 @@
+        return new KeyedCircuitBreakerMapping(perHost, perMethod, perPath, factory);
+    }
+
+    boolean isPerHost() {
+        return perHost;
+    }
+
+    boolean isPerMethod() {
+        return perMethod;
+    }
+
+    boolean isPerPath() {
+        return perPath;
+    }
```
Could remove?

haithamgabr replied:
Done. Thanks.

haithamgabr commented on core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerFactory.java:
```diff
@@ -0,0 +1,35 @@
+ * to a {@link CircuitBreaker}.
+ */
+@FunctionalInterface
+interface CircuitBreakerFactory {
```
Yes thanks a lot for spotting this. Done.

ikhoon reviewed (approved):
Great work! @haithamgabr
